mdiag/mdiag.sh

```diff
@@ -832,24 +830,19 @@ function _ungraboutput {
 
 function _json_strings_arrayify {
 	local a=("$@")
-	a=("${a[@]//\\/\\\\}") # this fixes vim syntax highlighting -> "
-	a=("${a[@]//\"/\\\"}")
-	a=("${a[@]//	/\\t}")
-	a=("${a[@]//$_lf/\\r}")
-	a=("${a[@]/#/\"}")
-	a=("${a[@]/%/\",}")
-	if [ ${#a[@]} -gt 0 ]; then
-		a[$(( ${#a[@]} - 1 ))]="${a[$(( ${#a[@]} - 1 ))]%,}"
-	fi
+	for i in ${!a[@]}; do
+	   a[$i]=$(_json_stringify "${a[$i]}")
+	   if [ ${#a[@]} -gt 0 -a $i -lt $(( ${#a[@]} - 1 )) ]; then
+	      a[$i]="${a[$i]},"
+	   fi
+	done
 	echo -n "[ ${a[@]} ]"
 }
 
 function _json_stringify {
-	local s="$*"
-	s="${s//\\/\\\\}" # this fixes vim syntax highlighting -> "
-	s="${s//\"/\\\"}"
-	s="${s//	/\\t}"
-	s="${s//$_lf/\\r}"
+	# The use of echo for an additional trailing newline is intentional
+	local s="$(echo -e "$*" | sed -e 's/\\/\\\\/g' -e 's/\"/\\\"/g' \
```

Comment:
Thanks for pointing out how forking affects performance in a hot path. It's been a while since i've played properly with shell scripting :) I've refactored this script now to minimise forking. The new stringify function results in slower execution on my machine, going from 2:40 to 2:55, however with the previous forking method it was 3:20 with twice the amount of CPU used compared to 2.0.4.